metronome/networking/src/io/iohk/metronome/networking/ConnectionHandler.scala

```diff
@@ -0,0 +1,215 @@
+package io.iohk.metronome.networking
+
+import cats.effect.{Concurrent, ContextShift, Resource}
+import cats.effect.concurrent.{Deferred, TryableDeferred}
+import io.iohk.metronome.networking.RemoteConnectionManager.{
+  MessageReceived,
+  UnexpectedConnectionError,
+  withCancelToken
+}
+import monix.catnap.ConcurrentQueue
+import monix.execution.atomic.AtomicInt
+import monix.tail.Iterant
+import cats.implicits._
+import cats.effect.implicits._
+import io.iohk.metronome.networking.ConnectionHandler.HandledConnection
+import io.iohk.metronome.networking.EncryptedConnectionProvider.ConnectionError
+
+import java.net.InetSocketAddress
+
+class ConnectionHandler[F[_]: Concurrent, K, M](
+    connectionQueue: ConcurrentQueue[F, HandledConnection[F, K, M]],
+    connectionsRegister: ConnectionsRegister[F, K, M],
+    messageQueue: ConcurrentQueue[F, MessageReceived[K, M]],
+    cancelToken: TryableDeferred[F, Unit],
+    connectionFinishCallback: HandledConnection[F, K, M] => F[Unit]
+) {
+
+  private val numberOfRunningConnections = AtomicInt(0)
+
+  private def closeAndDeregisterConnection(
+      handledConnection: HandledConnection[F, K, M]
+  ): F[Unit] = {
+    for {
+      _ <- Concurrent[F].delay(numberOfRunningConnections.decrement())
+      _ <- handledConnection.close()
```

Comment:
my assumption is that `close` should not fail be idempotent (at least that is the case in scalanet), but I agree deregister should come first otherwise we can return connection which is about to be closed in `getAcquired` connections